postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py

```
@@ -13,23 +17,40 @@ def download(self, local_dir='/tmp/', force=False):
 
         self.download_manager.open(
             'osmmftp.os.uk', self._username(), self._password())
-        self.download_manager.ftp_client.cwd(self._source_dir())
 
-        return self.download_manager.download_all_if_needed('./*_csv.zip',
-                                                            local_dir,
-                                                            force)
+        try:
+            self.ftp_client().cwd(self._source_dir())
+            return self.download_manager.download_all_if_needed('./*_csv.zip',
+                                                                local_dir,
+                                                                force)
+        except ftplib.error_perm:
+            logging.warning('FTP error - looking for local files')
+            # Ordnance Survey very kindly remove the files from our
+            # FTP area 21 days after they are first put there, despite
+            # there not being an update available for maybe another 2 months.
+            # So if we try to cd to our order dir
+            # and it's not there, we'll end up here.
+            # So, we then look for a load of csv.zip files in local_dir
+            # If there are csv.zip files locally, return them
+            # If not, look on s3.
+            # If they're on s3, download them and return the local paths.
+            # Otherwise FAIL
+            return self.local_files(local_dir) or self.files_from_s3(local_dir)
+
+    def ftp_client(self):
+        return self.download_manager.ftp_client
 
     def _username(self):
         username = os.environ.get('OS_FTP_USERNAME')
         if not username:
-            print 'OS_FTP_USERNAME not set!'
+            logging.warning('OS_FTP_USERNAME not set!')
```

For this and OS_FTP_PASSWORD, do we want to raise an exception instead?